### PR TITLE
PersistentDict: Use special marker struct for missing entries

### LINKED ISSUE #53584: ScopedValue is allocating when accessed
I have looked into using scoped values for some temporary arrays to avoid allocations in parallel tasks. However, it seems scoped values are allocating when accessed, whereas with tls it can be avoided.  This is unfortunate, since gc in parallel tasks can be a performance problem.

```
using .Threads
using BenchmarkTools

@noinline function tlsfun()
    tlsvec = get!(() -> [0], task_local_storage(), :myvec)::Vector{Int}
    tlsvec[1] += 1
    return nothing
end


const dynvec = ScopedValue([0])

@noinline function dynfun()
    dvec = dynvec[]
    dvec[1] += 1
    return nothing
end


function tlsrun()
    @sync for _ in 1:nthreads()
        @spawn for _ in 1:100000; tlsfun(); end
    end
end


function dynrun()
    @sync for _ in 1:nthreads()
        @with dynvec=>[0] @spawn for _ in 1:100000; dynfun(); end
    end
end


@btime tlsrun()
@btime dynrun()
versioninfo()
```

output:

```
  2.326 ms (202 allocations: 21.03 KiB)
  8.238 ms (2400274 allocations: 36.64 MiB)

Julia Version 1.12.0-DEV.121
Commit bc2212cc0e* (2024-03-04 01:20 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 24 × AMD Ryzen Threadripper PRO 5945WX 12-Cores
  WORD_SIZE: 64
  LLVM: libLLVM-16.0.6 (ORCJIT, znver3)
Threads: 24 default, 0 interactive, 12 GC (on 24 virtual cores)
Environment:
  JULIA_EDITOR = emacs -nw

```

### PR DESCRIPTION
**Pull request by topolarity**
This avoids allocating a `Tuple{T}` on each `get(dict::PersistentDict{K,V}, ...)` call

The other alternative here would be to force `KeyValue.get` to be inlined at these call-sites, so that SROA can get rid of the temporary but I assume @keno had good reason for trying to avoid that inlining.

Resolves #53584:

```julia
julia> const dynvec = ScopedValue([0])
julia> @noinline function dynfun()
           dvec = dynvec[]
           dvec[1] += 1
           return nothing
       end
julia> foo() = @with dynvec=>[0] for _ in 1:1000_000; dynfun(); end

julia> @allocated foo()
208
```

### CODE REVIEW

File: base/dict.jl [hunk 1]
```diff
@@ -978,53 +978,57 @@ end
 
 eltype(::PersistentDict{K,V}) where {K,V} = Pair{K,V}
 
+# XXX: This type is not legal to store in a PersistentDict
+struct NoEntry end
+const noentry = NoEntry()
```
Comment (vchuravy) [inline base/dict.jl:983]: Elsewhere we use a `gensym` symbol for this purpose.

**Review by vchuravy — COMMENTED** [review]
I wonder if Keno was avoiding the inlining so that the optimizations for KeyValue can locate them. For maximum optimization potential we would want to do the optimizations after inlining, but then we would want to run a second inlining pass for these pseudo-intriniscs.

**Review by Keno — CHANGES_REQUESTED** [review]
This doesn't work as is - the change violates the OptimizedGenerics interface.

**Comment by topolarity** [comment]
Oh, good point:
> The core optimization semantically allowed for the compiler is: 
>    get(set(x, key, val), key) -> (val,)

Maybe the ABI improvement is the right way to go about this.

**Comment by Keno** [comment]
The ABI improvement for this was a remaining to-do. I had started on it, but never got around to it. It's the correct fix.

**Pull request closed without merging**